Fastscape coupling review

Djneu opened the pull request:
Hi Anne,
Here's the latest coupled version, and thank you for reviewing it!

Commit:
Add fastscape files

Commit:
Add correct indexing for aspect->fastscape

Commit:
Add repetitions to nx and ny

Commit:
Allow code other than fastscape functions to run on multiple processors

Commit:
Change cell iterator and get velocity solutions

Commit:
Added  data tables for velocity components.

Commit:
Clean up code

Commit:
Fix send/receive for vector

Commit:
Add additional refinement for fastscape

Commit:
Allow 2D use with fastscape.

Commit:
Clean up code

Commit:
More comments and indent.

Commit:
Starting advection

Commit:
Add extra boundary nodes for 3d fastscape and timer.

Commit:
Add extra nodes to 2D models. This messes up the diffusion test.

Commit:
Fix ghost node indexing.

Commit:
Clean up code, output text to log, and only initialize fastscape at first timestep

Commit:
Change fastscape seed to a user defined variable.

Commit:
Track VTK file needed to write fastscape visualization into aspect folder.

Commit:
Add first pass at periodic boundaries.

Commit:
Clean up indexing.

Commit:
Update periodic boundaries

Commit:
Add restart to fastscape.

Commit:
Clean up code and indent.

Commit:
Fix 2D indexing with ghost nodes.

Commit:
Edit cmake to find fastscape library through FASTSCAPE_DIR

Commit:
Minor text changes

Commit:
Fix fastscape linking in cmake

Commit:
Small changes and add prm file to test fastscape

Commit:
Run indent and remove unnecessary includes

Commit:
Minor comment on cmakelists

Commit:
Add in variable for surface resolution

Commit:
Replace all numx/numy with nx/ny.

Commit:
Changes to get FastScape to work with non-uniform surface meshes.

Commit:
Add in more timers.

Commit:
Run indent

Commit:
Optimize transferring variables to single processor

Commit:
Add input parameters to use marine component.

Commit:
Change MPI_Send to MPI_Ssend

Commit:
Add first part of stratigraphy

Commit:
Add mass flux into model through ghost nodes.

Commit:
indent

Commit:
Minor change to variable name

Update variable name and make folder output only set if using stratigraphy

Commit:
Default slopep to zero

Default all values in slopep to zero in case there is an issue when copying from fastscape.

Commit:
Turn on initial seed

Commit:
Comment out folder output function

Commit:
Comment out folder output

Commit:
Fix averaging across nodes in 2d

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+    void
+    FastScape<dim>::initialize ()
+    {
+      const GeometryModel::Box<dim> *geometry
```
Please use the Plugin functionality to check whether a certain plugin used, e.g.Plugins::plugin_type_matches<const GeometryModel::EllipsoidalChunk<dim>> (this->get_geometry_model())

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+      restart = this->get_parameters().resume_computation;
+      restart_step = 0;
+
+      //second is for maximum coordiantes, first for minimum.
```
Please use the same format for your comments throughout in terms of spaces and capitalization, e.g. `// Second`.

Also, reverse comment and coordiantes->coordinates

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+      for (unsigned int i=0; i<dim; ++i)
+        {
+          grid_extent[i].first = 0;
+          grid_extent[i].second = geometry->get_extents()[i];
```
What if the origin of the box is displaced, i.e. its left front bottom corner doesn't lie at (0,0,0)? See geometry->get_origin()

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+        }
+
+      //TODO: There has to be a better type to use to get this.
+      const std::pair<int, int > repetitions = geometry->get_repetitions();
```
I'd just return the int array, as is used to store the repetitions in the box.cc itself.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+      //TODO: There has to be a better type to use to get this.
+      const std::pair<int, int > repetitions = geometry->get_repetitions();
+      int x_repetitions = repetitions.first;
```
make both int's const if you don't alter them below.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+      //Set nx and dx, as these will be the same regardless of dimension.
+      nx = 3+std::pow(2,surface_resolution+additional_refinement)*x_repetitions;
+      dx = grid_extent[0].second/(nx-3);
+      x_extent = geometry->get_extents()[0]+2*dx;
```
Do you need to call geometry->get_extents again? I think you can use grid_extent[0].second.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+      if (dim == 2)
+        {
+          dy = dx;
+          y_extent = round(y_extent_2d/dy)*dy+2*dy;             //x_extent; //grid_extent[0].second*3+2*dy;
```
remove commented out stuff

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+          ny = 1+y_extent/dy;
+        }
+
+      if (dim == 3)
```
instead of two if's, can you make it `if(dim ==2) ... else ...`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+          ny = 3+std::pow(2,surface_resolution+additional_refinement)*y_repetitions;
+          dy = grid_extent[1].second/(ny-3);
+          table_intervals[1] = ny-3;
+          y_extent = geometry->get_extents()[1]+2*dy;
```
use grid_extent instead of geometry->get_extents

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                                                             const std::set<types::boundary_id> &boundary_ids) const
+    {
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "Fastscape plugin");
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
```
As every boundary could potentially be deformed, I would check in `initialize` whether this plugin is called for top, as it only makes sense for the surface.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+    {
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "Fastscape plugin");
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      const bool top_boundary = boundary_ids.find(relevant_boundary) == boundary_ids.begin();
```
The surface boundary id does not necessarily need to be the first entry in the boundary_ids set. There could be other boundaries in the set. If `boundary_ids.find(relevant_boundary) == boundary_ids.end()` you could just `return` out of this function, because it only makes sense for the surface. Or if you do the check I suggested above, aspect will throw any way.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+      TimerOutput::Scope timer_section(this->get_computing_timer(), "Fastscape plugin");
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      const bool top_boundary = boundary_ids.find(relevant_boundary) == boundary_ids.begin();
+      int current_timestep = this->get_timestep_number ();
```
Make variables `const` when you don't intend to change them. Please check everywhere.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+        }
+
+      //We only want to run fastscape if there was a change in time, and if we're at the top boundary.
+      if (a_dt > 0 && top_boundary)
```
With the changes suggested above, the check for `top_boundary` can be removed.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                    if ( cell->face(face_no)->boundary_id() != relevant_boundary)
+                      continue;
+
+                    std::vector<Tensor<1,dim> > vel( face_corners.size() );
```
rm spaces around `face_corners.size()`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+                        /*
+                         * If we're in 2D, we want to take the values and apply them to every row of X points.
+                         * TODO: I need to test that 2D still works after changing how we store these variables.
```
Still need to test?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                         */
+                        if (dim == 2)
+                          {
+                            for (int ys=0; ys<ny; ys++)
```
please use `unsigned int` when looping over int's that are always positive

Djneu replied:
I've made this change for most loops, but it looks like FastScape won't take nx, ny, and array_size as an unsigned int. This ends up giving lots of warnings when setting the loop int to unsigned.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+                            double index = (indy-1)*nx+indx;
+
+                            temporary_variables[1].push_back(index-1);
```
compared to the 2D case, the indices are swapped, is that correct? If yes, please swap the 2 lines.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                            temporary_variables[0].push_back(vertex(dim-1));   //z component
+
+                            for (unsigned int i=0; i<dim; ++i)
+                              temporary_variables[i+2].push_back(vel[corner][i]*year_in_seconds);
```
should you always apply `year_in_seconds`, or only when `this->convert_output_to_years()`?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+              //Create variables for output directory and restart file
+              std::string filename;
+              filename = this->get_output_directory();
```
You can do this on one line and make it `const`. Also, technically it's a dirname and not a filename :)

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+              const std::string restart_filename = filename + "fastscape_h_restart.txt";
+              const std::string restart_step_filename = filename + "fastscape_steps_restart.txt";
+
+              //Initialize all FastScape variables.
```
Doubles are automatically initialized to zero. If you check that for the double arrays this is also true, then you don't have to manually set all entries to zero here.

Djneu replied:
This is odd, so the way I had it doesn't initialize them to zero. I found that by changing it from
 
`std::unique_ptr<double[]> h (new double[array_size]);`   
to
`std::unique_ptr<double[]> h (new double[array_size]());`   

It will initialize everything except the very last element to zero, and the last element seems to have a random number. I can initialize them as array_size+1, and then all the important ones are set to zero, and fastscape doesn't seem to care if the array is too large.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                  vx[temporary_variables[1][i]]= temporary_variables[2][i];
+                  vz[temporary_variables[1][i]]= temporary_variables[dim+1][i];
+
+                  if (dim == 2 )
```
if (2d) else

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+              for (unsigned int p=1; p<Utilities::MPI::n_mpi_processes(this->get_mpi_communicator()); ++p)
+                {
+
```
rm empty line

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+                  for (unsigned int i=0; i<temporary_variables.size(); i++)
+                    MPI_Recv(&temporary_variables[i][0], incoming_size, MPI_DOUBLE, p, 42, this->get_mpi_communicator(), &status);
+
```
rm empty line

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                      vx[temporary_variables[1][i]] = temporary_variables[2][i];
+                      vz[temporary_variables[1][i]] = temporary_variables[dim+1][i];
+
+                      if (dim == 2 )
```
if (2d) else

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+              if (current_timestep == 1 || restart)
+                {
+
```
We try to stick to 2 empty lines between plugin functions and 1 empty line between statements within these functions. No empty lines are needed after or before brackets.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+              if (current_timestep == 1 || restart)
+                {
+
+                  this->get_pcout() <<"   Initializing FastScape... "<< (1+surface_resolution+additional_refinement)  <<
```
one space before and after `<<`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                  //If we are restarting from a checkpoint, load h values for fastscape so we don't lose resolution.
+                  if (restart)
+                    {
+                      this->get_pcout() <<"      Loading FastScape restart file... "<<std::endl;
```
Check that you're using the same number of spaces before text printed into the log file.

Djneu replied:
In this case, I used extra spaces to have it seem as a subset of the previous fastscape text saying it was being initialized. Would it be better to keep all text the same?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                  if (restart)
+                    {
+                      this->get_pcout() <<"      Loading FastScape restart file... "<<std::endl;
+                      restart = false;
```
please move this line to the end of the if statement body. It's confusing to see it right after `if(restart)`.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                      in_step.open(restart_step_filename.c_str());
+                      if (in)
+                        {
+
```
rm

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                  std::ofstream out_h (restart_filename.c_str());
+                  std::ofstream out_step (restart_step_filename.c_str());
+
+                  out_step<<(istep+restart_step)<<std::endl;
```
spaces around `<<`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                  out_step<<(istep+restart_step)<<std::endl;
+
+                  for (int i=0; i<=array_size; i++)
+                    out_h<<h[i]<<std::endl;
```
idem

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+               */
+              fastscape_copy_slope_(slopep.get());
+
+              //Now we set the ghost nodes at the left and right boundaries.
```
Please add a description of the ghost nodes (how many, on which boundaries, what data you need there etc)

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                      //If its the first timestep add in initial slope. If we have no flux,
+                      //set the ghost node to the node next to it.
+                      slope = left_flux/kdd;
+                      h[index_left-1] = h[index_left] + slope*2*dx;
```
Please explain in a comment why you need `slope*2*dx`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                      //Because the corner nodes always show a slope of zero, this will update them according to
+                      //the nodes next to them.
+                      if (j == 0)
+                        slope = left_flux/kdd - std::tan(slopep[index_left+nx]*boost::math::double_constants::pi/180);
```
pi is in `numbers::PI`

anne-glerum replied:
For clarity, and to avoid issues with mixing int's and double's in formula's, I try to write `180.` when I mean a double.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+                  //If we set the boundaries as periodic, then reset any values to the
+                  //nodes on the opposite side.
+                  if (left == 0 && right == 0)
```
Note to self: where are left and right declared/set?

Djneu replied:
Left and right are set in the fastscape boundary conditions part of the prm file, where 1 is fixed and 0 is open.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                    }
+                }
+
+              //Now do the same for the top and bottom ghost nodes.
```
Please explain the ghost node logic below

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+
+                  if (bottom == 0 && top == 0)
+                    {
+                      double side = index_bot;
```
When declaring new variables, please add a comment saying what they are.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+              fastscape_set_h_(h.get());
+
+              //Initialize first time step, and update steps.
+              int visualization_step = istep+restart_step;
```
Add comment for `visualization_step`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                 * TODO: The frequency in this needs to be the same as the total timesteps fastscape will
+                 * run for, need to figure out how to work this in better.
+                 */
+                if (use_strat && current_timestep == 1)
```
What if `use_strat` and `current_timestep`>1?

Djneu replied:
I think stratigraphy will need to be revisited, but I believe the reasoning behind this was that you only need to call strati at the start and then it will handle all the visualization aspects. If you don't call it, then the VTK function needs to be called every time you want to make a visualization.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                  fastscape_destroy_();
+                }
+
+              //Find out our velocities from the change in height.
```
As you don't use `V` before this point, please move it's declaration to here. Otherwise the reader has forgotten what it was :)

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+            }
+          else
+            {
+              TimerOutput::Scope timer_section(this->get_computing_timer(), "Fastscape 1 proc");
```
Add a comment to describe what you're doing here

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+              size_idx[d] = table_intervals[d]+1;
+            }
+
+          Table<dim,double> data_table;
```
Add comments for these new variables

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                }
+            }
+
+          if (dim == 3)
```
if (2d) else

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                          idx[0] = j;
+
+                          //We only care about the surface
+                          if (k==1)
```
So there are several points in the z direction? Why not `k==0`?

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+    template <int dim>
+    void FastScape<dim>::declare_parameters(ParameterHandler &prm)
+    {
+
```
rm

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+    void FastScape<dim>::declare_parameters(ParameterHandler &prm)
+    {
+
+      prm.declare_entry ("End time",
```
Why are you declaring `End time` again? This is already declared. There is no need to declare it again if you mean to parse it in parse_parameters.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                            "Number of steps per ASPECT timestep");
+          prm.declare_entry("Maximum timestep", "10e3",
+                            Patterns::Double(0),
+                            "Maximum timestep for fastscape.");
```
fastscape -> FastScape everywhere in text

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                            "Vertical exaggeration for fastscape's VTK file.");
+          prm.declare_entry("Additional fastscape refinement", "0",
+                            Patterns::Integer(),
+                            "Refinement level expected at surface to determine"
```
From the name and description it's not clear how this differs from `Surface resolution`

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                            Patterns::Integer(),
+                            "The difference between the lowest and highest resolution level at surface. So if three resolution "
+                            "levels are expected, this would be set to 2.");
+          prm.declare_entry ("Use marine parameters", "false",
```
Is this to use marine parameters or to use the whole marine functionality? If the latter, update the name to reflect this.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                            "Total number of steps you expect in the FastScape model, only used if stratigraphy is turned on.");
+          prm.declare_entry("Number of horizons", "1",
+                            Patterns::Integer(),
+                            "Number of horizons to track and visualize in FastScape..");
```
double dots

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                              Patterns::Double(),
+                              "Exponent to determine the distribution from the SPL to neighbor nodes, with"
+                              "10 being steepest decent and 1 being more varied.");
+            prm.declare_entry("Bedrock deposition coefficient", "-1",
```
do these negative default values make sense?

Djneu replied:
FastScape takes values < 0 to imply that the value isn't used, although maybe using zero would work as well and be less confusing.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+                              "e-folding depth for the exponential of the shale porosity law.");
+            prm.declare_entry("Sand-shale ratio", "0.5",
+                              Patterns::Double(),
+                              "ratio of sand to shale for material leaving continent.");
```
ratio->Ratio

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1054 @@
+      }
+      prm.leave_subsection ();
+
+      prm.enter_subsection ("Mesh refinement");
```
Again, no need to redefine the parameter.

anne-glerum commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,175 @@
+#ifdef __cplusplus
+extern"C" {
+#endif
+//Functions to initialize FastScape
```
Do these functions have to be outside the FastScape plugin?

Djneu replied:
It looks like they can be moved into either the aspect or mesh deformation namespace, but I get the error that they're out of scope if I put them into the FastScape class.

anne-glerum commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,175 @@
+
+      private:
+        /**
+         * A function object representing the mesh deformation.
```
What does this comment refer to?

anne-glerum commented on include/aspect/mesh_deformation/fastscape.h:
```diff
@@ -0,0 +1,175 @@
+         * A function object representing the mesh deformation.
+         */
+        int nstep;
+        int bc;
```
Please add a comment to each parameter

anne-glerum commented on include/aspect/geometry_model/box.h:
```diff
@@ -71,6 +71,13 @@ namespace aspect
+         * Return a point that denotes the repetitions of the box in each dimension
+         * of the domain.
+         */
+        std::pair<int, int >
```
I'd just return the int array `unsigned int repetitions[dim]`

anne-glerum reviewed (changes requested):
Hi Derek, this is amazing work! I've gone through it once now and noted down everything I saw in terms of formatting, comments and the easy logic. When you've updated the plugin to address my comments, I'll go through it again and really focus on the logic of setting grid values, ghost nodes etc. Thanks for your work!!

anne-glerum commented:
I'm installing this branch now. FYI I get a lot of compiler warnings of this kind:
![Screenshot 2020-06-02 at 15 24 02](https://user-images.githubusercontent.com/7631624/83525799-cb177680-a4e5-11ea-8bf5-9993efe29f32.png)
Please make everything that should/could be an integer, an integer.

anne-glerum commented:
I also had to change the pi version to numbers::PI (which is the one generally used in ASPECT), and the linker fails with dealii 9.1.1, not with 9.0.1.

Djneu commented:
Hi Anne,

Thank you for catching the compiler warnings, it looks like I initialized the side and index's when setting the ghost nodes as double for some reason, but they should always be integers. Is there a setting you compile with to see those warnings?

I'm not sure about the linker error, are you installing on a cluster or your local computer? I'm using the release version for deal 9.1.1 on glogin and haven't run into a linker issue yet.

anne-glerum commented:
> I'm not sure about the linker error, are you installing on a cluster or your local computer? I'm using the release version for deal 9.1.1 on glogin and haven't run into a linker issue yet

I installed on my mac. I just built aspect dev with 9.1.1 without linker errors. Maybe you just need to update, the errors did not seem to be in your code.

Edit: I think I had 2 dealii versions active at the time of installation that may have corrupted the install. 

> Is there a setting you compile with to see those warnings?

I think the default compiler settings on my mac are different. But yes, you can set compiler flags, see here for example: https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html

anne-glerum commented on cookbooks/fastscape_eroding_plateau.prm:
```diff
@@ -0,0 +1,152 @@
+
+# X boundaries are periodic, and the rest are tangential.
+subsection Boundary velocity model
+  set Tangential velocity boundary indicators = bottom, front, back
```
What about the boundary conditions for the left and right boundary?

Commit:
Switch branch

Commit:
Address some comments, create VTK folder in aspect.

Commit:
Address some formatting, move where h_old is set.

Commit:
Address more comments

Commit:
Redo indexing for ghost nodes.

Commit:
Add switch for ghost nodes.

Commit:
Update prm file and run astyle.

Commit:
remove extra line

Commit:
Comment format in fastscape.h

Commit:
Prm file comment format.

Djneu commented:
Hi Anne,
I've hopefully addressed all the comments and added a few things to the code. Thanks again for taking the time to review this!

Commit:
Fix initial topography advection error and astyle

Commit:
Fix isolines plugin. There are still some 'todo's in the '.cc' file. And as the dealii in the docker is also older than requested, I haven't generated the output yet.

MFraters committed:
Initial layout of isolines plugin.

gassmoeller committed:
Modify a function to compute reference viscosity

lhy11009 committed:
add feature to isolines.cc:
1. background composition
2. comments

modification to isolines.cc:
1. change some typos
2. following the 'volume_fraction' method in MaterialUtilities class to rewrite the part for composition

add tests:
1. isolines_slab_2D
2. isolines_slab_2D_background
3. isolines_slab_3D

MFraters committed:
Change isolines to isosurfaces.

MFraters committed:
add changelog isosurfaces mesh refinement plugin.

MFraters committed:
adress comments Timo.

MFraters committed:
simplify tests

MFraters committed:
add min, min-1, min-2, max, max-1, etc functionality

MFraters committed:
add documentation, clean and add functionality to isolines code

MFraters committed:
create and use is_defect_correction helper function.

Commit:
Fix assert for multiple conditions

Fix the assert so that it doesn't go off if a separate mesh deformation indicator is called on the top boundary.

Commit:
Change fastscape base

Before the top of the model was considered the top of fastscape. This caused issues with the ghost node basement on initialization, so now FastScape is always set with 0 as the base.

EstherHeck committed:
add defect correction stokes to list of allowed solvers for strain dependent and elasticity

anne-glerum committed:
Clean up

Commit:
Fix bad merge and reapply later commits.

gassmoeller committed:
Add changelog

gassmoeller committed:
Fix broken tests

Commit:
Round index values before placing them.

Commit:
Add time intervals for sediment rain.

Commit:
add cout statement.

Commit:
move stabilization to interface

Commit:
Remove free surface stabilization parameter

Commit:
Make fastscape visualizations at teh same time as ASPECT.

Commit:
Minor fix to VTK output.

gassmoeller committed:
Change remove particles

gassmoeller committed:
Allow particle initialization at boundary

Commit:
Update FastScape VTK file to output additional fields.

Commit:
Adjust ghost nodes so they are fully periodic in ASPECT.

Commit:
Move ghost nodes to separate function

bangerth committed:
Improvements to the van Keken benchmark section of the manual.

anne-glerum committed:
Update particle handler copy

Commit:
Hopefully fix restarts for VTK output.

gassmoeller committed:
Connect signals only once.

gassmoeller committed:
Checkpoint backup particles

ljhwang committed:
Adding 2 publications Barrioneuvo and Bredow

anne-glerum committed:
Indent

tjhei committed:
add link to future plans

Commit:
Add cout for writing vtk, and change ConstraintMatrix to AffineConstraints.

MFraters committed:
Merge pull request #3985 from tjhei/future_plans

add link to future plans

anne-glerum committed:
Add assert max_visc >= min_visc

naliboff committed:
Merge pull request #3925 from EstherHeck/update_allowed_solvers_in_strain_dependent

add defect correction to list of allowed solvers for strain dependent

naliboff committed:
Merge pull request #3986 from anne-glerum/min_max_viscosity

Add assert max_visc >= min_visc

anne-glerum committed:
Remove backup from checkpointing

anne-glerum committed:
Update test

tjhei committed:
add file to gitignore

naliboff committed:
Merge pull request #3990 from tjhei/gitignore1

add file to gitignore

bangerth committed:
Merge pull request #3984 from ljhwang/citations_february

Adding 2 publications Barrioneuvo and Bredow

bangerth committed:
Initialize particles in one go.

EstherHeck committed:
added elastic stress contribution to stress and shear stress vis postprocessor

EstherHeck committed:
adjusted error message for time stepping function

MFraters committed:
Change computation of strain rate invariant in visualization postprocessor.

MFraters committed:
apply changes from tester.

ljhwang committed:
Adding 4 publications properly this time

bangerth committed:
Slightly improve the mark-up in the manual.

tjhei committed:
Merge pull request #3998 from bangerth/ref

Slightly improve the mark-up in the manual.

bangerth committed:
Use the 'cite' package for the manual.

The package makes sure that references are listed not as [4,2,8,5] but as
[2,4,5,8]. As a side effect, it also replaces [1,3,5,6,7,8] by [1,3,5-8].

bangerth committed:
Various improvements to the manual.

tjhei committed:
Merge pull request #3993 from EstherHeck/change-error-message-in-time-stepping-function

adjusted error message for time stepping function

anne-glerum committed:
Address comments

MFraters committed:
Apply suggestions from Timo's code review on isosurfaces plugin

Co-authored-by: Timo Heister <timo.heister@gmail.com>

MFraters committed:
Further adressing Timo's comments.

EstherHeck committed:
changed all compositional stress fields from stress_xx etc to ve_stress_xx etc in .cc , .h , .prm and python files

MFraters committed:
Merge pull request #4001 from EstherHeck/add-elasticity-to-stress-postprocessor

added elastic stress contribution to stress and shear stress vis pp

bangerth committed:
More improvements to formatting in the manual.

bangerth committed:
Style changes.

EstherHeck committed:
updated tests

EstherHeck committed:
renamed stress_xx etc in changes from PR#4001

anne-glerum committed:
Move assert

anne-glerum committed:
Add test

anne-glerum committed:
Remove unused parameter

anne-glerum committed:
Add first steps for new users

anne-glerum committed:
Update test

gassmoeller committed:
Make sure visualization names are unique

MFraters committed:
Merge pull request #4000 from bangerth/manual

Various improvements to the manual.

anne-glerum committed:
Address comments

MFraters committed:
Merge pull request #3997 from ljhwang/citations_MarchNew

Adding 4 publications properly this time

MFraters committed:
Merge pull request #4002 from EstherHeck/rename_elastic_stresses_compositional_fields

renamed compositional stress fields

gassmoeller committed:
Merge pull request #3932 from anne-glerum/particles_plus_timestepping

Particles plus repeated timestepping

bobmyhill committed:
Merge pull request #4007 from anne-glerum/first_steps_troubleshooting

Add first steps for new users

EstherHeck committed:
add Rene and Esther to doc file

EstherHeck committed:
enabled stress averaging without using fixed lastic timestep

EstherHeck committed:
documentation of and suggestions for the parameter

EstherHeck committed:
updated test

bobmyhill committed:
Merge pull request #4004 from bangerth/manual-3

More improvements to formatting in the manual.

bobmyhill committed:
Merge pull request #4005 from bangerth/style

Style changes.

tjhei committed:
Merge pull request #3999 from bangerth/cite

Use the 'cite' package for the manual.

bangerth committed:
Add another paper based on ASPECT.

tjhei committed:
Merge pull request #4012 from bangerth/citing

Add another paper based on ASPECT.

EstherHeck committed:
update test result

bangerth committed:
Avoid the use of a couple of functions that were just deprecated in deal.II.

tjhei committed:
Merge pull request #4013 from bangerth/deprecated

Avoid the use of a couple of functions that were just deprecated in deal.II

tjhei committed:
GMG: reuse AlignedVector

Reusing the AlignedVector in the GMG code skips allocating memory over
and over. Prompted by https://github.com/dealii/dealii/issues/12008

alarshi committed:
Included constraints for interpolate output into advection field.

alarshi committed:
Corrected the implementation.

tjhei committed:
delay exit with failure in parallel

When an exception (like solver failure) is triggered, all but rank 0
immeditately return in main() without printing anything. Some MPI
implementations detect abnormal termination and abort the whole job.
This could happen before rank 0 produces an error message.
Avoid this by sleeping before exiting.

bangerth committed:
Merge pull request #4014 from tjhei/gmg_speedup

GMG: reuse AlignedVector

bangerth committed:
Merge pull request #4017 from tjhei/wait_exit

delay exit with failure in parallel

gassmoeller committed:
Merge pull request #3992 from bangerth/particles-2

Initialize particles in one go.

tjhei committed:
add missing header file

gassmoeller committed:
Merge pull request #4020 from tjhei/add_missing_include

add missing header file

gassmoeller committed:
Fix spherical shell restart. Add test.

tjhei committed:
Merge pull request #4022 from gassmoeller/fix_spherical_shell_restart

Fix spherical shell restart. Add test.

gassmoeller committed:
Merge pull request #4011 from EstherHeck/elasticity_stabilization_parameter

Elasticity stabilization parameter

gassmoeller committed:
Make indent

gassmoeller committed:
Merge pull request #4006 from anne-glerum/spherical_gaussian

Fix spherical gaussian perturbation initial temperature

gassmoeller committed:
Merge pull request #3995 from MFraters/change_strain_rate_visualization_postprocessor

Change computation of strain rate invariant in visualization postprocessor.

gassmoeller committed:
Update test

gassmoeller committed:
Merge pull request #3980 from bangerth/manual-2

Improvements to the van Keken benchmark section of the manual.

tjhei committed:
Merge pull request #4009 from gassmoeller/unique_visualization_names

Make sure visualization names are unique

alarshi committed:
Modified the test results to use correct temperature bcs.

gassmoeller committed:
Merge pull request #4015 from alarshi/bcs_with_prescribed_temperature

Included constraints for interpolate output into advection field.

gassmoeller committed:
Remove unnecessary particle backups

ljhwang committed:
Add one new citation and update Heckenbach

gassmoeller committed:
Merge pull request #4025 from ljhwang/master

Add one new citation, Faccenna, and update Heckenbach

MFraters committed:
Fix split line problem for cmake version 2.8.12.

MFraters committed:
Merge comments Bob

Co-authored-by: Bob Myhill <myhill.bob@gmail.com>

anne-glerum committed:
Add test

gassmoeller committed:
Fix probability density particle generator

bangerth committed:
Merge pull request #4027 from gassmoeller/fix_probability_density_generator

Fix probability density particle generator

anne-glerum committed:
Change particle output to gnuplot

gassmoeller committed:
Merge pull request #4026 from MFraters/fix_cmake_for_version_2.8.12

Fix split line problem for cmake version 2.8.12.

MFraters committed:
Merge pull request #4024 from gassmoeller/remove_duplicated_backup

Remove unnecessary particle backups

bangerth committed:
Add a publication using ASPECT.

gassmoeller committed:
Work around removed exception

gassmoeller committed:
Merge pull request #4030 from bangerth/citing

Add a publication using ASPECT.

tjhei committed:
Merge pull request #4032 from gassmoeller/fix_deal_ii_dev

Work around removed exception

gassmoeller committed:
Address comments

gassmoeller committed:
Implement and test closest boundary id

gassmoeller committed:
Update test

gassmoeller committed:
Indent

bangerth committed:
Adjust some code to our usual style.

jdannberg committed:
Merge pull request #3981 from gassmoeller/particle_boundary_condition

Allow particle initialization with Dirichlet boundary condition

bobmyhill committed:
Merge pull request #4034 from bangerth/style

Adjust some code to our usual style.

gassmoeller committed:
Add a fallback interpolator to bilinear least squares

jdannberg committed:
fix lateral viscosity variations

jdannberg committed:
add changelog and update tests

gassmoeller committed:
Remove duplicated calls to initialize

gassmoeller committed:
Merge pull request #4042 from jdannberg/steinberger_lateral_viscosity

fix lateral viscosity variations

gassmoeller committed:
Update test results

tjhei committed:
Merge pull request #4043 from gassmoeller/remove_duplicate_initialize

Remove duplicated calls to initialize

bangerth committed:
Add an explanatory sentence to the 'compositing' material model.

tjhei committed:
add dependency info to readme

tjhei committed:
address comment

bangerth committed:
Avoid the use of Threads::Thread.

Use std::thread instead since the deal.II class is deprecated.

bangerth committed:
Avoid the use of TriaAccessor::number_of_children().

This function is deprecated. Use TriaAccessor::n_children() instead.

bangerth committed:
Avoid a deprecated version of MappingQCache::initialize().

bangerth committed:
Avoid the use of deprecated functions for iterating over local vector elements.

bangerth committed:
Merge pull request #4045 from tjhei/readme-cmake

add dependency info to readme

bangerth committed:
Merge pull request #4036 from gassmoeller/add_fallback_for_particle_interpolation

Add a fallback interpolator to bilinear least squares

MFraters committed:
Merge pull request #4048 from bangerth/depr-neighbors

Avoid the use of TriaAccessor::number_of_children().

MFraters committed:
Merge pull request #4049 from bangerth/depr-initialize

Avoid a deprecated version of MappingQCache::initialize().

tjhei committed:
Merge pull request #4047 from bangerth/threads

Avoid the use of Threads::Thread.

gassmoeller committed:
Merge pull request #4050 from bangerth/depr-local-size

Avoid the use of deprecated functions for iterating over local vector elements.

anne-glerum committed:
Merge pull request #4044 from bangerth/compositing

Add an explanatory sentence to the 'compositing' material model.

anne-glerum committed:
Add elastic stress to surface stress

anne-glerum committed:
Update names

gassmoeller committed:
Update FEFieldFunction templates

gassmoeller committed:
Merge pull request #4055 from gassmoeller/make_compile_with_deal_dev

Update FEFieldFunction templates

MFraters committed:
Apply suggestions from code review from Anne

Co-authored-by: anne-glerum <a.c.glerum@uu.nl>

MFraters committed:
Adressing review comments from Anne.

tjhei committed:
update parameters

tjhei committed:
Merge pull request #3725 from MFraters/isolines

Isosurface mesh refinment plugin

gassmoeller committed:
Merge pull request #4057 from tjhei/maintenance_june2021

update parameters

gassmoeller committed:
Fix deprecation warning

gassmoeller committed:
Merge pull request #4059 from gassmoeller/fix_deprecation

Fix deprecation warning

anne-glerum committed:
Update output names

anne-glerum committed:
Update output names in test

gassmoeller committed:
compatibility patch

gassmoeller committed:
Merge pull request #4053 from anne-glerum/elastic_surface_stress

Add elastic stresses to surface stress output

bangerth committed:
Add a missing assertion.

tjhei committed:
Merge pull request #4065 from bangerth/assertion

Add a missing assertion.

tjhei committed:
update copyright years

tjhei committed:
fix bunge cookbook

fix:
1. missing slash (Ascii data file                                                                                       </home/heister/aspect-git/cookbooks/bunge_et_al_mantle_convectionvisc_depth_a.txt>
    not found!
2. add POINTS header to viscosity files, reported as
    Could not successfully read in the file header of the ascii data file
3. add column information to parse column names / count correctly

tjhei committed:
add Assert to StructuredData

When the number of components can not be parsed (no column names
specified) and not specified in the constructor, we would crash with
bad:alloc. Instead, add an Assert.

Found while working on #4071

tjhei committed:
fix some doxygen comment formatting

zjiaqi2018 committed:
fix deprecation warnings

gassmoeller committed:
Merge pull request #4073 from tjhei/doxygen_fixes

fix some doxygen comment formatting

gassmoeller committed:
Merge pull request #4072 from tjhei/structured_assert

add Assert to StructuredData

gassmoeller committed:
Merge pull request #4071 from tjhei/fix_bunge_cookbook

fix bunge cookbook

gassmoeller committed:
Merge pull request #4074 from zjiaqi2018/fix_deprecation_warnings

Fix matrix-free deprecation warnings

gassmoeller committed:
Merge pull request #4070 from tjhei/copyright

update copyright years

gassmoeller committed:
Clarify comment

gassmoeller committed:
Update a few tests

gassmoeller committed:
Update test results

gassmoeller committed:
Update changelog

tjhei committed:
Merge pull request #4063 from gassmoeller/make_compatible_with_particle_vector

Make particle removal compatible with latest deal.II particle structure

gassmoeller committed:
Merge pull request #3681 from gassmoeller/compute_reference_viscosity

Modify a function to compute reference viscosity

gassmoeller committed:
Update docker container versions

gassmoeller committed:
Merge pull request #4075 from gassmoeller/update_docker_instructions

Update docker container versions

Commit:
Fix initial topo and adiabatic conditions error.

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      restart = this->get_parameters().resume_computation;
+      restart_step = 0;
+
+      // second is for maximum coordinates, first for minimum.
```
Note that when the origin is not at zero, the maximum coordinates are origin+extents

anne-glerum commented on source/mesh_deformation/fastscape.cc:
```diff
@@ -0,0 +1,1237 @@
+      const unsigned int x_repetitions = geometry->get_repetitions(0);
+      const unsigned int y_repetitions = geometry->get_repetitions(1);
+
+      // Set nx and dx, as these will be the same regardless of dimension.
```
This logic needs to be explained in the comment.
